File: app/src/main/java/com/bresiu/codechallenge/repository/mappers/Mapper.java
package com.bresiu.codechallenge.repository.mappers;

import java.util.ArrayList;
import java.util.List;

public abstract class Mapper {
	public static <S, T> List<S> mapList(List<T> objectList, ObjectMapper<S, T> objectMapper) {
		List<S> resultList = new ArrayList<>(objectList.size());
		for (T object : objectList) {
			resultList.add(objectMapper.map(object));
		}
		return resultList;
	}

	public static <S, T> S mapObject(T object, ObjectMapper<S, T> objectMapper) {
		return objectMapper.map(object);
	}

	public interface ObjectMapper<S, T> {
		S map(T object);
	}
}

File: app/src/main/java/com/bresiu/codechallenge/repository/mappers/Mapper.kt
package com.bresiu.codechallenge.repository.mappers

import java.util.*

object Mapper {
  fun <S, T> mapList(objectList: List<T>, objectMapper: ObjectMapper<S, T>): List<S> {
    val resultList = ArrayList<S>(objectList.size)
    for (mappingItem in objectList) {
      resultList.add(objectMapper.map(mappingItem))
    }
    return resultList
  }

  fun <S, T> unfoldList(objectList: List<T>, objectMapper: ObjectMapper<List<S>, T>): List<S> {
    val resultList = ArrayList<S>()
    for (listItem in objectList) {
      resultList.addAll(objectMapper.map(listItem))
    }
    return resultList
  }

  fun <S, T> mapObject(mappingItem: T, objectMapper: ObjectMapper<S, T>): S {
    return objectMapper.map(mappingItem)
  }

  interface ObjectMapper<S, T> {
    fun map(mappingItem: T): S
  }
}
